File: REFORGE-OS-3.0.0/Bobbys-Workshop--3.0.0/BobbysWorkshop-Installer-v1.2.0/server/flash-progress-server.js
#!/usr/bin/env node

const WebSocket = require('ws');
const express = require('express');
const cors = require('cors');

const app = express();
app.use(cors());
app.use(express.json());

const PORT = Number(process.env.FLASH_PROGRESS_PORT || 3002);

console.log('[flash-progress-server] Starting standalone WS server.');
console.log('[flash-progress-server] NOTE: The main backend (server/index.js) already provides /ws/flash-progress on port 3001.');
console.log(`[flash-progress-server] This standalone server listens on port ${PORT} to avoid conflicts.`);

const wss = new WebSocket.Server({ port: PORT });

class FlashProgressManager {
  constructor() {
    this.clients = new Set();
    this.activeJobs = new Map();
  }

  broadcast(message) {
    const data = JSON.stringify({
      ...message,
      timestamp: Date.now()
    });

    console.log('[WS] Broadcasting:', message.type, message.jobId || '');

    this.clients.forEach(client => {
      if (client.readyState === WebSocket.OPEN) {
        try {
          client.send(data);
        } catch (err) {
          console.error('[WS] Send error:', err);
        }
      }
    });
  }

  flashStarted(jobId, deviceId, deviceName, totalBytes) {
    this.activeJobs.set(jobId, {
      jobId,
      deviceId,
      deviceName,
      status: 'running',
      startedAt: Date.now()
    });

    this.broadcast({
      type: 'flash_started',
      jobId,
      deviceId,
      deviceName,
      stage: 'Initializing',
      totalBytes
    });
  }

  sendProgress(jobId, deviceId, progress, stage, bytesTransferred, 
               totalBytes, transferSpeed, eta) {
    const job = this.activeJobs.get(jobId);
    if (job) {
      job.progress = progress;
      job.stage = stage;
    }

    this.broadcast({
      type: 'flash_progress',
      jobId,
      deviceId,
      progress,
      stage,
      bytesTransferred,
      totalBytes,
      transferSpeed,
      estimatedTimeRemaining: eta
    });
  }

  flashCompleted(jobId, deviceId) {
    const job = this.activeJobs.get(jobId);
    if (job) {
      job.status = 'completed';
    }

    this.broadcast({
      type: 'flash_completed',
      jobId,
      deviceId
    });
  }

  flashFailed(jobId, deviceId, error) {
    const job = this.activeJobs.get(jobId);
    if (job) {
      job.status = 'failed';
      job.error = error;
    }

    this.broadcast({
      type: 'flash_failed',
      jobId,
      deviceId,
      error
    });
  }

  flashPaused(jobId, deviceId) {
    const job = this.activeJobs.get(jobId);
    if (job) {
      job.status = 'paused';
    }

    this.broadcast({
      type: 'flash_paused',
      jobId,
      deviceId
    });
  }

  flashResumed(jobId, deviceId) {
    const job = this.activeJobs.get(jobId);
    if (job) {
      job.status = 'running';
    }

    this.broadcast({
      type: 'flash_resumed',
      jobId,
      deviceId
    });
  }
}

const manager = new FlashProgressManager();

wss.on('connection', (ws) => {
  manager.clients.add(ws);
  console.log(`[WS] Client connected. Total: ${manager.clients.size}`);

  ws.on('message', (data) => {
    try {
      const message = JSON.parse(data.toString());
      
      if (message.type === 'ping') {
        ws.send(JSON.stringify({
          type: 'pong',
          timestamp: Date.now()
        }));
      }
    } catch (err) {
      console.error('[WS] Message parse error:', err);
    }
  });

  ws.on('close', () => {
    manager.clients.delete(ws);
    console.log(`[WS] Client disconnected. Total: ${manager.clients.size}`);
  });

  ws.on('error', (err) => {
    console.error('[WS] Error:', err);
  });
});

async function performFlash(jobId, deviceId, deviceName, partition, imageSize) {
  console.log(`[FLASH] Starting job ${jobId} for device ${deviceId}`);
  
  const totalBytes = imageSize || 4294967296;
  const chunkSize = 10485760;
  const updateInterval = 500;
  
  manager.flashStarted(jobId, deviceId, deviceName, totalBytes);
  
  let bytesTransferred = 0;
  let startTime = Date.now();
  
  return new Promise((resolve) => {
    const interval = setInterval(() => {
      bytesTransferred += chunkSize;
      
      if (bytesTransferred >= totalBytes) {
        clearInterval(interval);
        manager.flashCompleted(jobId, deviceId);
        console.log(`[FLASH] Completed job ${jobId}`);
        resolve();
        return;
      }
      
      const progress = (bytesTransferred / totalBytes) * 100;
      const elapsed = (Date.now() - startTime) / 1000;
      const transferSpeed = bytesTransferred / elapsed;
      const remaining = totalBytes - bytesTransferred;
      const eta = remaining / transferSpeed;
      
      let stage = 'Writing system partition';
      if (progress < 5) stage = 'Initializing flash operation';
      else if (progress < 10) stage = 'Verifying device connection';
      else if (progress < 20) stage = 'Unlocking bootloader';
      else if (progress < 30) stage = 'Erasing existing partition';
      else if (progress < 90) stage = `Flashing ${partition} partition`;
      else if (progress < 95) stage = 'Verifying flash integrity';
      else stage = 'Finalizing and rebooting';
      
      manager.sendProgress(
        jobId, deviceId, progress, stage,
        bytesTransferred, totalBytes, transferSpeed, eta
      );
    }, updateInterval);
  });
}

app.post('/api/flash/start', async (req, res) => {
  const jobId = `job_${Date.now()}_${Math.random().toString(36).substr(2, 9)}`;
  const { deviceId, deviceName, partition, imageSize } = req.body;
  
  console.log(`[API] Starting flash job: ${jobId}`);
  console.log(`[API] Device: ${deviceId} (${deviceName})`);
  console.log(`[API] Partition: ${partition}`);
  console.log(`[API] Image size: ${imageSize} bytes`);
  
  performFlash(jobId, deviceId, deviceName || deviceId, partition || 'system', imageSize);
  
  res.json({ 
    success: true,
    jobId, 
    status: 'started',
    deviceId,
    partition
  });
});

app.get('/api/flash/jobs', (req, res) => {
  const jobs = Array.from(manager.activeJobs.values());
  res.json({ jobs });
});

app.get('/api/flash/job/:jobId', (req, res) => {
  const job = manager.activeJobs.get(req.params.jobId);
  if (job) {
    res.json({ job });
  } else {
    res.status(404).json({ error: 'Job not found' });
  }
});

app.post('/api/flash/demo', async (req, res) => {
  const devices = [
    { id: 'PIXEL6_001', name: 'Google Pixel 6', partition: 'system', size: 3221225472 },
    { id: 'SAMSUNG_S21_002', name: 'Samsung Galaxy S21', partition: 'boot', size: 2147483648 },
    { id: 'ONEPLUS_9_003', name: 'OnePlus 9 Pro', partition: 'vendor', size: 1073741824 }
  ];

  const jobs = devices.map(device => {
    const jobId = `job_${Date.now()}_${Math.random().toString(36).substr(2, 9)}`;
    
    setTimeout(() => {
      performFlash(jobId, device.id, device.name, device.partition, device.size);
    }, Math.random() * 2000);
    
    return {
      jobId,
      deviceId: device.id,
      deviceName: device.name,
      partition: device.partition
    };
  });

  res.json({ 
    success: true,
    message: 'Demo flash operations started',
    jobs 
  });
});

app.get('/health', (req, res) => {
  res.json({ 
    status: 'ok',
    websocket: {
      port: 3001,
      connected_clients: manager.clients.size
    },
    active_jobs: manager.activeJobs.size
  });
});

const HTTP_PORT = process.env.PORT || 3000;

app.listen(HTTP_PORT, '0.0.0.0', () => {
  console.log('');
  console.log('╔═══════════════════════════════════════════════════════════╗');
  console.log('║  Bobby\'s World - Flash Progress WebSocket Server         ║');
  console.log('╠═══════════════════════════════════════════════════════════╣');
  console.log(`║  HTTP API:       http://localhost:${HTTP_PORT}                     ║`);
  console.log('║  WebSocket:      ws://localhost:3001/flash-progress       ║');
  console.log('╠═══════════════════════════════════════════════════════════╣');
  console.log('║  Endpoints:                                               ║');
  console.log('║    POST /api/flash/start    - Start flash job            ║');
  console.log('║    POST /api/flash/demo     - Start demo jobs (3 devices)║');
  console.log('║    GET  /api/flash/jobs     - List active jobs           ║');
  console.log('║    GET  /health             - Server health check        ║');
  console.log('╚═══════════════════════════════════════════════════════════╝');
  console.log('');
  console.log('[SERVER] Ready to accept connections...');
  console.log('');
});

process.on('SIGINT', () => {
  console.log('\n[SERVER] Shutting down gracefully...');
  wss.clients.forEach(client => {
    client.close();
  });
  wss.close(() => {
    console.log('[WS] WebSocket server closed');
    process.exit(0);
  });
});

File: REFORGE-OS-3.0.0/Bobbys-Workshop--3.0.0/BobbysWorkshop-Installer-v1.2.0/server/flash_progress_server.py
#!/usr/bin/env python3
"""
Bobby's World - Flash Progress WebSocket Server (Python/FastAPI)

A WebSocket server for live flash operation progress updates.
"""

from fastapi import FastAPI, WebSocket, WebSocketDisconnect
from fastapi.middleware.cors import CORSMiddleware
from pydantic import BaseModel
import asyncio
import json
import time
from typing import Dict, List, Optional
from datetime import datetime
import uvicorn

app = FastAPI(title="Bobby's World Flash Progress Server")

app.add_middleware(
    CORSMiddleware,
    allow_origins=["*"],
    allow_credentials=True,
    allow_methods=["*"],
    allow_headers=["*"],
)


class FlashJobRequest(BaseModel):
    deviceId: str
    deviceName: Optional[str] = None
    partition: Optional[str] = "system"
    imageSize: Optional[int] = 4294967296


class FlashProgressManager:
    def __init__(self):
        self.active_connections: List[WebSocket] = []
        self.active_jobs: Dict[str, dict] = {}

    async def connect(self, websocket: WebSocket):
        await websocket.accept()
        self.active_connections.append(websocket)
        print(f"[WS] Client connected. Total: {len(self.active_connections)}")

    def disconnect(self, websocket: WebSocket):
        if websocket in self.active_connections:
            self.active_connections.remove(websocket)
            print(f"[WS] Client disconnected. Total: {len(self.active_connections)}")

    async def broadcast(self, message: dict):
        """Broadcast message to all connected clients"""
        message["timestamp"] = int(time.time() * 1000)
        data = json.dumps(message)
        
        print(f"[WS] Broadcasting: {message.get('type')} {message.get('jobId', '')}")
        
        disconnected = []
        for connection in self.active_connections:
            try:
                await connection.send_text(data)
            except Exception as e:
                print(f"[WS] Send error: {e}")
                disconnected.append(connection)

        for connection in disconnected:
            self.disconnect(connection)

    async def flash_started(self, job_id: str, device_id: str, device_name: str, total_bytes: int):
        """Notify that a flash operation has started"""
        self.active_jobs[job_id] = {
            "jobId": job_id,
            "deviceId": device_id,
            "deviceName": device_name,
            "status": "running",
            "startedAt": int(time.time() * 1000)
        }
        
        await self.broadcast({
            "type": "flash_started",
            "jobId": job_id,
            "deviceId": device_id,
            "deviceName": device_name,
            "stage": "Initializing",
            "totalBytes": total_bytes
        })

    async def send_progress(self, job_id: str, device_id: str, progress: float,
                           stage: str, bytes_transferred: int, total_bytes: int,
                           transfer_speed: float, eta: float):
        """Send progress update for a specific job"""
        if job_id in self.active_jobs:
            self.active_jobs[job_id]["progress"] = progress
            self.active_jobs[job_id]["stage"] = stage
        
        await self.broadcast({
            "type": "flash_progress",
            "jobId": job_id,
            "deviceId": device_id,
            "progress": progress,
            "stage": stage,
            "bytesTransferred": bytes_transferred,
            "totalBytes": total_bytes,
            "transferSpeed": transfer_speed,
            "estimatedTimeRemaining": eta
        })

    async def flash_completed(self, job_id: str, device_id: str):
        """Notify that a flash operation completed successfully"""
        if job_id in self.active_jobs:
            self.active_jobs[job_id]["status"] = "completed"
        
        await self.broadcast({
            "type": "flash_completed",
            "jobId": job_id,
            "deviceId": device_id
        })

    async def flash_failed(self, job_id: str, device_id: str, error: str):
        """Notify that a flash operation failed"""
        if job_id in self.active_jobs:
            self.active_jobs[job_id]["status"] = "failed"
            self.active_jobs[job_id]["error"] = error
        
        await self.broadcast({
            "type": "flash_failed",
            "jobId": job_id,
            "deviceId": device_id,
            "error": error
        })

    async def flash_paused(self, job_id: str, device_id: str):
        """Notify that a flash operation was paused"""
        if job_id in self.active_jobs:
            self.active_jobs[job_id]["status"] = "paused"
        
        await self.broadcast({
            "type": "flash_paused",
            "jobId": job_id,
            "deviceId": device_id
        })

    async def flash_resumed(self, job_id: str, device_id: str):
        """Notify that a flash operation was resumed"""
        if job_id in self.active_jobs:
            self.active_jobs[job_id]["status"] = "running"
        
        await self.broadcast({
            "type": "flash_resumed",
            "jobId": job_id,
            "deviceId": device_id
        })


manager = FlashProgressManager()


@app.websocket("/flash-progress")
async def flash_progress_endpoint(websocket: WebSocket):
    await manager.connect(websocket)
    try:
        while True:
            data = await websocket.receive_text()
            message = json.loads(data)

            if message.get("type") == "ping":
                await websocket.send_json({
                    "type": "pong",
                    "timestamp": int(time.time() * 1000)
                })
    except WebSocketDisconnect:
        manager.disconnect(websocket)
    except Exception as e:
        print(f"[WS] Error: {e}")
        manager.disconnect(websocket)


async def perform_flash(job_id: str, device_id: str, device_name: str, 
                       partition: str, image_size: int):
    """
    Simulate a flash operation with progress updates
    """
    print(f"[FLASH] Starting job {job_id} for device {device_id}")
    
    total_bytes = image_size
    chunk_size = 10485760  # 10MB
    update_interval = 0.5  # seconds
    
    await manager.flash_started(job_id, device_id, device_name, total_bytes)
    
    bytes_transferred = 0
    start_time = time.time()
    
    try:
        while bytes_transferred < total_bytes:
            await asyncio.sleep(update_interval)
            bytes_transferred += chunk_size
            
            if bytes_transferred > total_bytes:
                bytes_transferred = total_bytes
            
            progress = (bytes_transferred / total_bytes) * 100
            elapsed = time.time() - start_time
            transfer_speed = bytes_transferred / elapsed if elapsed > 0 else 0
            remaining = total_bytes - bytes_transferred
            eta = remaining / transfer_speed if transfer_speed > 0 else 0
            
            # Determine stage based on progress
            if progress < 5:
                stage = "Initializing flash operation"
            elif progress < 10:
                stage = "Verifying device connection"
            elif progress < 20:
                stage = "Unlocking bootloader"
            elif progress < 30:
                stage = "Erasing existing partition"
            elif progress < 90:
                stage = f"Flashing {partition} partition"
            elif progress < 95:
                stage = "Verifying flash integrity"
            else:
                stage = "Finalizing and rebooting"
            
            await manager.send_progress(
                job_id, device_id, progress, stage,
                bytes_transferred, total_bytes, transfer_speed, eta
            )
        
        await manager.flash_completed(job_id, device_id)
        print(f"[FLASH] Completed job {job_id}")
        
    except Exception as e:
        error_msg = str(e)
        await manager.flash_failed(job_id, device_id, error_msg)
        print(f"[FLASH] Failed job {job_id}: {error_msg}")


@app.post("/api/flash/start")
async def start_flash(request: FlashJobRequest):
    """Start a flash operation"""
    job_id = f"job_{int(time.time())}_{hex(hash(time.time()))[2:10]}"
    device_name = request.deviceName or request.deviceId
    
    print(f"[API] Starting flash job: {job_id}")
    print(f"[API] Device: {request.deviceId} ({device_name})")
    print(f"[API] Partition: {request.partition}")
    print(f"[API] Image size: {request.imageSize} bytes")
    
    asyncio.create_task(perform_flash(
        job_id, request.deviceId, device_name,
        request.partition, request.imageSize
    ))
    
    return {
        "success": True,
        "jobId": job_id,
        "status": "started",
        "deviceId": request.deviceId,
        "partition": request.partition
    }


@app.get("/api/flash/jobs")
async def get_jobs():
    """Get all active jobs"""
    return {"jobs": list(manager.active_jobs.values())}


@app.get("/api/flash/job/{job_id}")
async def get_job(job_id: str):
    """Get specific job details"""
    job = manager.active_jobs.get(job_id)
    if job:
        return {"job": job}
    return {"error": "Job not found"}, 404


@app.post("/api/flash/demo")
async def start_demo():
    """Start demo flash operations for multiple devices"""
    devices = [
        {
            "id": "PIXEL6_001",
            "name": "Google Pixel 6",
            "partition": "system",
            "size": 3221225472
        },
        {
            "id": "SAMSUNG_S21_002",
            "name": "Samsung Galaxy S21",
            "partition": "boot",
            "size": 2147483648
        },
        {
            "id": "ONEPLUS_9_003",
            "name": "OnePlus 9 Pro",
            "partition": "vendor",
            "size": 1073741824
        }
    ]
    
    jobs = []
    for i, device in enumerate(devices):
        job_id = f"job_{int(time.time())}_{i}_{hex(hash(time.time() + i))[2:10]}"
        
        delay = i * 0.5
        asyncio.create_task(
            delayed_flash(delay, job_id, device["id"], device["name"],
                        device["partition"], device["size"])
        )
        
        jobs.append({
            "jobId": job_id,
            "deviceId": device["id"],
            "deviceName": device["name"],
            "partition": device["partition"]
        })
    
    return {
        "success": True,
        "message": "Demo flash operations started",
        "jobs": jobs
    }


async def delayed_flash(delay: float, *args):
    """Start flash after delay"""
    await asyncio.sleep(delay)
    await perform_flash(*args)


@app.get("/health")
async def health_check():
    """Health check endpoint"""
    return {
        "status": "ok",
        "websocket": {
            "port": 8000,
            "connected_clients": len(manager.active_connections)
        },
        "active_jobs": len(manager.active_jobs)
    }


if __name__ == "__main__":
    print("")
    print("╔═══════════════════════════════════════════════════════════╗")
    print("║  Bobby's World - Flash Progress WebSocket Server         ║")
    print("╠═══════════════════════════════════════════════════════════╣")
    print("║  HTTP API:       http://localhost:8000                   ║")
    print("║  WebSocket:      ws://localhost:8000/flash-progress      ║")
    print("╠═══════════════════════════════════════════════════════════╣")
    print("║  Endpoints:                                               ║")
    print("║    POST /api/flash/start    - Start flash job            ║")
    print("║    POST /api/flash/demo     - Start demo jobs (3 devices)║")
    print("║    GET  /api/flash/jobs     - List active jobs           ║")
    print("║    GET  /health             - Server health check        ║")
    print("╚═══════════════════════════════════════════════════════════╝")
    print("")
    print("[SERVER] Starting server...")
    print("")
    
    uvicorn.run(app, host="0.0.0.0", port=8000)
